<a href="https://colab.research.google.com/github/kpipinis/M1_Assignment_Stock_Price_Prediction_with_LLM./blob/main/M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [2]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


/tmp/ipykernel_2462/1433581267.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.128755  128.671053  125.473652  127.051113  1.120920e+08
std      30.301251   30.577591   30.020452   30.329240  5.602586e+07
min      54.163700   55.160698   51.324796   55.059309  3.519590e+07
25%     112.264143  113.753656  110.812201  112.443578  7.636470e+07
50%     132.485161  133.613784  130.532798  132.307403  9.493580e+07
75%     147.635815  148.798284  145.698868  147.005556  1.296327e+08
max     177.939697  178.848900  175.114320  178.545835  4.265100e+08


In [3]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())


/tmp/ipykernel_2462/730687086.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.128755  128.671053  125.473652  127.051113  1.120920e+08
std      30.301251   30.577591   30.020452   30.329240  5.602586e+07
min      54.163700   55.160698   51.324796   55.059309  3.519590e+07
25%     112.264143  113.753656  110.812201  112.443578  7.636470e+07
50%     132.485161  133.613784  130.532798  132.307403  9.493580e+07
75%     147.635815  148.798284  145.698868  147.005556  1.296327e+08
max     177.939697  178.848900  175.114320  178.545835  4.265100e+08


In [4]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



/tmp/ipykernel_2462/3699561310.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.128755  128.671053  125.473652  127.051113  1.120920e+08
std      30.301251   30.577591   30.020452   30.329240  5.602586e+07
min      54.163700   55.160698   51.324796   55.059309  3.519590e+07
25%     112.264143  113.753656  110.812201  112.443578  7.636470e+07
50%     132.485161  133.613784  130.532798  132.307403  9.493580e+07
75%     147.635815  148.798284  145.698868  147.005556  1.296327e+08
max     177.939697  178.848900  175.114320  178.545835  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

In [ ]:
Here are some features that might help improve a stock price prediction model:

1.  **Historical Prices**: Beyond just closing prices, consider open, high, low, and adjusted close prices.
2.  **Volume**: Trading volume can indicate the strength of price movements.
3.  **Moving Averages**: Simple Moving Averages (SMA) and Exponential Moving Averages (EMA) over various periods (e.g., 5-day, 10-day, 20-day, 50-day, 200-day) can smooth price data and identify trends.
4.  **Technical Indicators**: A wide range of indicators can provide insights into momentum, volatility, and overbought/oversold conditions:
    *   **Relative Strength Index (RSI)**: Measures the speed and change of price movements.
    *   **Moving Average Convergence Divergence (MACD)**: Reveals changes in the strength, direction, momentum, and duration of a trend.
    *   **Bollinger Bands**: Measure market volatility and identify overbought or oversold conditions.
    *   **Stochastic Oscillator**: Compares a closing price to its price range over a given time period.
    *   **Average True Range (ATR)**: Measures market volatility by decomposing the entire range of an asset price for that period.
    *   **On-Balance Volume (OBV)**: Relates volume to price change to detect price movements.
5.  **Volatility Measures**: Besides ATR, other measures like standard deviation of returns over different periods can capture market risk.
6.  **Lagged Features**: Including past values of prices, volumes, or indicators can help capture temporal dependencies.
7.  **Time-based Features**: Day of the week, month, quarter, year, or even specific market holidays can sometimes reveal patterns.
8.  **Economic Indicators**: Macroeconomic factors can influence overall market sentiment and stock performance:
    *   Interest rates (e.g., Fed Funds Rate)
    *   Inflation rates (e.g., CPI)
    *   Gross Domestic Product (GDP) growth
    *   Unemployment rates
    *   Consumer confidence index
    *   Producer Price Index (PPI)
9.  **Company-Specific Fundamentals**: For individual stocks, financial ratios and news can be highly relevant:
    *   Earnings per Share (EPS)
    *   Price-to-Earnings (P/E) ratio
    *   Revenue growth
    *   Debt-to-Equity ratio
    *   Analyst ratings and price targets
    *   News sentiment (from news articles, social media, earnings call transcripts)
10. **Inter-Market Analysis / Comparative Features**: Performance relative to broader market indices (e.g., S&P 500, NASDAQ) or industry peers.
11. **Option Data**: For more advanced models, implied volatility from options contracts can be a forward-looking indicator.

# New section

## Step 3.1: Model Building - Linear Regression

In [8]:
!pip install ta
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import ta
import pandas as pd # Ensure pandas is imported

# Assuming 'data' DataFrame is populated as described in the kernel state.

# Flatten multi-level columns from yfinance output for easier access
# The kernel state shows columns like ('Price', 'AAPL'), ('Close', 'AAPL'), etc.
if isinstance(data.columns, pd.MultiIndex):
    # Take the first level of the MultiIndex (e.g., 'Close', 'Price')
    data.columns = data.columns.get_level_values(0)

# 1. Feature Engineering

# Historical Prices (Open, High, Low, Close) and Volume are now directly accessible
# 'Close' often represents the adjusted close in yfinance data, especially with auto_adjust=True.

# Moving Averages
data['SMA_5'] = data['Close'].rolling(window=5).mean()
data['SMA_10'] = data['Close'].rolling(window=10).mean()
data['SMA_20'] = data['Close'].rolling(window=20).mean()
data['SMA_50'] = data['Close'].rolling(window=50).mean()

# Technical Indicators: RSI and MACD
data['RSI'] = ta.momentum.RSIIndicator(data['Close'], window=14).rsi()
macd = ta.trend.MACD(data['Close'])
data['MACD_line'] = macd.macd()
data['MACD_signal'] = macd.macd_signal()

# Create the target variable: 'Tomorrow_Close'
# This is the 'Close' price of the next trading day
data['Tomorrow_Close'] = data['Close'].shift(-1)

# Drop rows with NaN values introduced by feature engineering and target variable creation
df_model = data.dropna()

# Define features (X) and target (y)
# Removed 'Price' from features as 'Close' is typically the adjusted price
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'SMA_5', 'SMA_10', 'SMA_20', 'SMA_50', 'RSI', 'MACD_line', 'MACD_signal']
X = df_model[features]
y = df_model['Tomorrow_Close']

# 2. Train-Test Split (70/30 chronological split)
split_index = int(len(df_model) * 0.7)

X_train, X_test = X[:split_index], X[split_index:]
y_train, y_test = y[:split_index], y[split_index:]

print(f"Training set size: {len(X_train)} samples")
print(f"Testing set size: {len(X_test)} samples")

# 3. Model Training (Linear Regression)
model = LinearRegression()
model.fit(X_train, y_train)

# 4. Model Evaluation
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"\nMean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R²): {r2:.2f}")

Training set size: 494 samples
Testing set size: 212 samples

Mean Squared Error (MSE): 11.42
R-squared (R²): 0.92


## Step 3.2: Model Building - Random Forest Classification

In [9]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Create the binary target variable: 1 if 'Close' increases tomorrow, 0 otherwise
data['Tomorrow_Direction'] = (data['Close'].shift(-1) > data['Close']).astype(int)

# Drop rows with NaN values introduced by feature engineering and new target variable
df_classifier_model = data.dropna()

# Define features (X) and the new target (y_direction)
# 'Tomorrow_Close' is no longer needed as a feature, but used to create 'Tomorrow_Direction'
features_clf = ['Open', 'High', 'Low', 'Close', 'Volume', 'SMA_5', 'SMA_10', 'SMA_20', 'SMA_50', 'RSI', 'MACD_line', 'MACD_signal']
X_clf = df_classifier_model[features_clf]
y_direction = df_classifier_model['Tomorrow_Direction']

# 2. Train-Test Split (70/30 chronological split)
split_index_clf = int(len(df_classifier_model) * 0.7)

X_train_clf, X_test_clf = X_clf[:split_index_clf], X_clf[split_index_clf:]
y_train_direction, y_test_direction = y_direction[:split_index_clf], y_direction[split_index_clf:]

print(f"Training set size (Classifier): {len(X_train_clf)} samples")
print(f"Testing set size (Classifier): {len(X_test_clf)} samples")

# 3. Model Training (Random Forest Classifier)
clf_model = RandomForestClassifier(n_estimators=100, random_state=42)
clf_model.fit(X_train_clf, y_train_direction)

# 4. Model Evaluation
y_pred_direction = clf_model.predict(X_test_clf)

accuracy = accuracy_score(y_test_direction, y_pred_direction)
conf_matrix = confusion_matrix(y_test_direction, y_pred_direction)
class_report = classification_report(y_test_direction, y_pred_direction)

print(f"\nAccuracy: {accuracy:.2f}")
print(f"\nConfusion Matrix:\n{conf_matrix}")
print(f"\nClassification Report:\n{class_report}")

# Feature Importance
feature_importances = pd.Series(clf_model.feature_importances_, index=features_clf).sort_values(ascending=False)
print(f"\nFeature Importance:\n{feature_importances}")

Training set size (Classifier): 494 samples
Testing set size (Classifier): 212 samples

Accuracy: 0.53

Confusion Matrix:
[[31 79]
 [21 81]]

Classification Report:
              precision    recall  f1-score   support

           0       0.60      0.28      0.38       110
           1       0.51      0.79      0.62       102

    accuracy                           0.53       212
   macro avg       0.55      0.54      0.50       212
weighted avg       0.55      0.53      0.50       212


Feature Importance:
Volume         0.098124
RSI            0.096371
High           0.086310
MACD_line      0.085771
MACD_signal    0.084779
SMA_50         0.084115
Open           0.082075
Close          0.080513
Low            0.080102
SMA_20         0.075621
SMA_10         0.075239
SMA_5          0.070980
dtype: float64


## Step 4: Model Evaluation & Prediction for 2023-01-01 to 2024-01-01

In [10]:
import yfinance as yf
import pandas as pd
import ta
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# 1. Load new data for the prediction period (2023-01-01 to 2024-01-01)
new_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Flatten multi-level columns from yfinance output
if isinstance(new_data.columns, pd.MultiIndex):
    new_data.columns = new_data.columns.get_level_values(0)

# 2. Engineer the same features on the new data
# Moving Averages
new_data['SMA_5'] = new_data['Close'].rolling(window=5).mean()
new_data['SMA_10'] = new_data['Close'].rolling(window=10).mean()
new_data['SMA_20'] = new_data['Close'].rolling(window=20).mean()
new_data['SMA_50'] = new_data['Close'].rolling(window=50).mean()

# Technical Indicators: RSI and MACD
new_data['RSI'] = ta.momentum.RSIIndicator(new_data['Close'], window=14).rsi()
macd_new = ta.trend.MACD(new_data['Close'])
new_data['MACD_line'] = macd_new.macd()
new_data['MACD_signal'] = macd_new.macd_signal()

# 3. Create the target variable 'Tomorrow_Direction' for evaluation
new_data['Tomorrow_Direction'] = (new_data['Close'].shift(-1) > new_data['Close']).astype(int)

# Drop rows with NaN values introduced by feature engineering and target variable creation
df_predict = new_data.dropna()

# Define features (X) and target (y) for the prediction period
X_predict = df_predict[features_clf] # Using the same features_clf from the previous model
y_true_predict = df_predict['Tomorrow_Direction']

print(f"Prediction set size: {len(X_predict)} samples")

# 4. Make predictions using the trained Random Forest Classifier (clf_model)
y_pred_new = clf_model.predict(X_predict)

# 5. Evaluate the performance on the new data
accuracy_new = accuracy_score(y_true_predict, y_pred_new)
conf_matrix_new = confusion_matrix(y_true_predict, y_pred_new)
class_report_new = classification_report(y_true_predict, y_pred_new)

print(f"\nAccuracy on new data (2023-2024): {accuracy_new:.2f}")
print(f"\nConfusion Matrix on new data:\n{conf_matrix_new}")
print(f"\nClassification Report on new data:\n{class_report_new}")

/tmp/ipykernel_2462/3631434720.py:7: FutureWarning: YF.download() has changed argument auto_adjust default to True
  new_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed


Prediction set size: 201 samples

Accuracy on new data (2023-2024): 0.46

Confusion Matrix on new data:
[[80 12]
 [96 13]]

Classification Report on new data:
              precision    recall  f1-score   support

           0       0.45      0.87      0.60        92
           1       0.52      0.12      0.19       109

    accuracy                           0.46       201
   macro avg       0.49      0.49      0.40       201
weighted avg       0.49      0.46      0.38       201



---

## Step 5: Strategy Discussion

### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a **low performance** with an accuracy of approximately **46.27%** on the new data (2023-2024). This is a significant drop compared to its performance on the original test set (53%) and suggests the model has poor generalization capabilities or that the market dynamics have changed.
- **Precision and Recall for Class 0 (No Increase)**: The model has a precision of 0.45 and a recall of 0.87 for predicting no price increase. This means when it predicts no increase, it's correct about 45% of the time, and it correctly identifies 87% of all instances of no price increase.
- **Precision and Recall for Class 1 (Increase)**: The model has a precision of 0.52 but a very low recall of 0.12 for predicting a price increase. This implies that while the model is correct about 52% of the time when it *does* predict an increase, it misses a large number of actual price increases (only catches 12% of them).

#### Potential Trading Strategies:

Given this performance, relying solely on this model for aggressive trading strategies is **highly risky**. However, we can infer some cautious approaches:

1.  **Extreme Caution / Risk Aversion Strategy (Focus on Avoiding Losses):**
    *   **Action:** Given the very low accuracy and recall for price increases, it would be prudent to **avoid buying or holding based on this model's 'increase' predictions**. The model is largely unable to reliably predict upward movements.
    *   **Risk Management:** The model's recall for 'no increase' (Class 0) is relatively high (0.87). This means when it predicts 'no increase', there's a good chance it's correct. Therefore, if the model predicts **'0' (no increase)**, it could be a strong signal to **avoid opening new long positions or even consider exiting existing ones to mitigate potential losses.** This leverages the model's slightly better ability to identify non-growth periods.

2.  **Short-Selling / Hedging Consideration (Highly Speculative):**
    *   **Focus:** Potentially use the 'no increase' signal for defensive or short-selling strategies.
    *   **Action:** If the model predicts a '0' (no increase), one *might* consider short-selling or hedging. However, given the low overall accuracy, this would still be very speculative and require significant external validation.

3.  **Model Improvement Focus:**
    *   **Focus:** The current model is not suitable for direct trading. The primary strategy should be to improve the model itself.
    *   **Action:** Explore more advanced feature engineering, different machine learning models (e.g., LSTMs for time series), hyperparameter tuning, or incorporating external data sources (e.g., news sentiment, economic indicators) to build a more robust and predictive model.
    *   **Risk Management:** Do not trade with this model until significant improvements in its predictive power, especially for 'Class 1' (price increase), are demonstrated.

**Overall Assessment:**
**The model's performance on the new data is very poor and suggests it is currently not useful for making profitable trading decisions.** Its inability to reliably predict upward movements (low recall for Class 1) means it would miss most buying opportunities. The moderately high recall for 'no increase' (Class 0) *could* be cautiously used as a signal to avoid risk, but even that has limited precision. Substantial work on feature engineering and model development is required before this model could be considered for any real-world trading strategy.